## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [3]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info

<bound method DataFrame.info of            first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  

In [5]:
df.dtypes

first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object

In [6]:
# What three item categories had the most entries?
df.category.value_counts().head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [7]:
# Create a new df with only data that the column "category" has "consumables"
consumables_df = df[df['category'] == 'consumables']
consumables_df.head()


first      last                    job                phone  \
1    Tiffany     Myers        Music therapist     201.442.4543x942   
3     Nathan     Baker  Accounting technician     827-788-8123x012   
4  Christina  Schwartz           Chiropractor         265-829-3643   
6      Brian     Moore                Barista  +1-893-400-2733x982   
7      Jesus    Wright     Theme park manager   605.947.9467x66513   

                                 email  client_id  order_id  order_date  \
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
6          moore.official@martinez.com      87213   8080280  2023-04-19   
7        wright.jesus.7579@meadows.com      46709   8946782  2023-02-14   

   order_week  order_year         item_id     category       subcategory  \
1          20        2023  XDA18116-89-4A  consumables              pens   
3          17        2023  ZMM00836-65-0C  consumables              pens   
4          21        2023  BZX55559-12-3X  consumables              misc   
6          16        2023  UAC50632-25-4E  consumables  kitchen supplies   
7           7        2023  UEU25316-01-1U  consumables              pens   

   unit_price  unit_cost  unit_weight  qty  line_number  
1       24.95      15.09         1.49   21            0  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1  
6      245.56     231.95         1.73   56            5  
7       20.99      16.97         5.12   22            4

In [8]:
# For the category with the most entries, which subcategory had the most entries?
consumables_df.subcategory.value_counts().head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [9]:
# Which five clients had the most entries in the data?
df.client_id.value_counts().head(5)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [10]:
test_value = df.client_id.value_counts().head(5)
test_value

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [11]:
df.client_id.value_counts().head(5).index

Index([33615, 66037, 46820, 38378, 24741], dtype='int64', name='client_id')

In [40]:
# Store the client ids of those top 5 clients in a list.
top_clients_list = df.client_id.value_counts().head(5).index.tolist()
top_clients_list

[33615, 66037, 46820, 38378, 24741]

In [13]:
top_client = int(df.client_id.value_counts().head(1).index[0])
top_client

33615

In [14]:
# How many total units (the qty column) did the client with the most entries order order?
highest_sell_df = df[df.client_id == top_client]
highest_sell_df.qty.sum()

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [15]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df["total_unit_cost"] = df.unit_price * df.qty
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  total_unit_cost  
0     1096.80     762.71         7.50  105            1        115164.00  
1       24.95      15.09         1.49   21            0           523.95  
2       13.52       7.86         1.68   39            6           527.28  
3       36.42      24.85         1.23   29            3          1056.18  
4      195.10     108.17        46.43   20            1          3902.00

In [20]:
df["total_weight"] = df.qty * df.unit_weight

In [21]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df["shipping_price"] = df.total_weight.apply(lambda x: 7 * x if x > 50 else 10 * x)  
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number  total_unit_cost  shipping_price  \
0         7.50  105            1        115164.00         5512.50   
1         1.49   21            0           523.95          312.90   
2         1.68   39            6           527.28          458.64   
3         1.23   29            3          1056.18          356.70   
4        46.43   20            1          3902.00         6500.20   

     total_price  total_weight  
0  125898.607500        787.50  
1     588.693625         31.29  
2     594.407400         65.52  
3    1167.314400         35.67  
4    4770.182750        928.60  

[5 rows x 22 columns]

In [23]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df["total_price"] = round((df.total_unit_cost + df.shipping_price) * 1.0925, 2)
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number  total_unit_cost  shipping_price  \
0         7.50  105            1        115164.00         5512.50   
1         1.49   21            0           523.95          312.90   
2         1.68   39            6           527.28          458.64   
3         1.23   29            3          1056.18          356.70   
4        46.43   20            1          3902.00         6500.20   

   total_price  total_weight  
0    131839.08        787.50  
1       914.26         31.29  
2      1077.12         65.52  
3      1543.57         35.67  
4     11364.40        928.60  

[5 rows x 22 columns]

In [24]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df["total_cost"] = df.unit_cost * df.qty + df.shipping_price
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number  total_unit_cost  shipping_price  total_price  total_weight  \
0            1        115164.00         5512.50    131839.08        787.50   
1            0           523.95          312.90       914.26         31.29   
2            6           527.28          458.64      1077.12         65.52   
3            3          1056.18          356.70      1543.57         35.67   
4            1          3902.00         6500.20     11364.40        928.60   

   total_cost  
0    85597.05  
1      629.79  
2      765.18  
3     1077.35  
4     8663.60  

[5 rows x 23 columns]

In [25]:
# Create a column for the profit of each line using line cost and line price
df["profit"] = df.total_price - df.total_cost
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

   total_unit_cost  shipping_price  total_price  total_weight  total_cost  \
0        115164.00         5512.50    131839.08        787.50    85597.05   
1           523.95          312.90       914.26         31.29      629.79   
2           527.28          458.64      1077.12         65.52      765.18   
3          1056.18          356.70      1543.57         35.67     1077.35   
4          3902.00         6500.20     11364.40        928.60     8663.60   

     profit  
0  46242.03  
1    284.47  
2    311.94  
3    466.22  
4   2700.80  

[5 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [30]:
list_of_test_order_ids = [2742071, 2173913, 6128929]

In [32]:
def total_of_specified_order_id(order_id):
    """
    Calculate the total amount for a specified order ID.

    This function simulates a database query by searching through a predefined
    list of orders and returning the total amount for the specified order ID.

    Args:
    order_id (int): The order ID to search for.

    Returns:
    float: The total amount for the specified order ID.
    If the order ID is not found, returns None.
    """

    return df.loc[df.order_id == order_id, "total_price"].sum()

In [35]:
# Check your work using the totals above
#df.loc[df.order_id == 2742071, "total_price"].sum()
round(total_of_specified_order_id(2742071), 2)

152811.89

In [37]:
for id in list_of_test_order_ids:
    print(f"Order {id} total: ${total_of_specified_order_id(id):,.2f}")

Order 2742071 total: $152,811.89
Order 2173913 total: $162,388.71
Order 6128929 total: $923,441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [42]:
def client_total_spend(client_id):
    temp_df = df[df.client_id == client_id]
    return temp_df["total_price"].sum()

In [43]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
for client in top_clients_list:
    print(f"Client Number: {client}: ${client_total_spend(client):,.2f}")


Client Number: 33615: $8,377,308.52
Client Number: 66037: $10,259,514.79
Client Number: 46820: $9,743,794.36
Client Number: 38378: $12,906,550.87
Client Number: 24741: $82,268,892.02


In [44]:
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

   total_unit_cost  shipping_price  total_price  total_weight  total_cost  \
0        115164.00         5512.50    131839.08        787.50    85597.05   
1           523.95          312.90       914.26         31.29      629.79   
2           527.28          458.64      1077.12         65.52      765.18   
3          1056.18          356.70      1543.57         35.67     1077.35   
4          3902.00         6500.20     11364.40        928.60     8663.60   

     profit  
0  46242.03  
1    284.47  
2    311.94  
3    466.22  
4   2700.80  

[5 rows x 24 columns]

In [ ]:
# Create a summary df with all the clients
summary_df = df[["client_id", "qty", "shipping_price", "total_price", "total_cost", "profit"]]
summary_df.head()

In [ ]:
summed_summary_df = summary_df.groupby("client_id", as_index=False).sum()
summed_summary_df

client_id    qty  shipping_price  total_price   total_cost      profit
0        10033   3153       139218.49    689274.95    493623.91   195651.04
1        10091  29711       561984.00   1825072.92   1230236.35   594836.57
2        10218  32559       483720.94   3683119.77   3029239.91   653879.86
3        10351   7615       240086.16   1712403.23   1246616.00   465787.23
4        10435  36967      1032718.36   8722369.83   6231649.51  2490720.32
..         ...    ...             ...          ...          ...         ...
994      99400  22892       540311.45   2418252.31   1704230.99   714021.32
995      99450   8890       360201.45   1063754.03    764042.43   299711.60
996      99736  63487      1259645.64   4558684.38   3302345.17  1256339.21
997      99973   5797       211914.88    983748.40    755545.89   228202.51
998      99984  48034      2951092.51  18704140.50  12167684.21  6536456.29

[999 rows x 6 columns]

In [51]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
top_five_summary_df = summed_summary_df[summed_summary_df.client_id.isin(top_clients_list)]
top_five_summary_df.reset_index(drop=True, inplace=True)
top_five_summary_df

client_id     qty  shipping_price  total_price   total_cost       profit
0      24741  239862      5126448.37  82268892.02  45688899.71  36579992.31
1      33615   64313      1828984.89   8377308.52   6175313.91   2201994.61
2      38378   73667      3429455.40  12906550.87   9634720.98   3271829.89
3      46820   75768      1601448.84   9743794.36   7007191.64   2736602.72
4      66037   43018      1395151.85  10259514.79   7004482.98   3255031.81

In [54]:
top_five_summary_df.sort_values(by= ["profit"], inplace=True)
top_five_summary_df

C:\Users\William Frescas\AppData\Local\Temp\ipykernel_33628\3878650869.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  top_five_summary_df.sort_values(by= ["profit"], inplace=True)


client_id     qty  shipping_price  total_price   total_cost       profit
1      33615   64313      1828984.89   8377308.52   6175313.91   2201994.61
3      46820   75768      1601448.84   9743794.36   7007191.64   2736602.72
4      66037   43018      1395151.85  10259514.79   7004482.98   3255031.81
2      38378   73667      3429455.40  12906550.87   9634720.98   3271829.89
0      24741  239862      5126448.37  82268892.02  45688899.71  36579992.31

In [55]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 
money_columns_list = ["shipping_price", "total_price", "total_cost", "profit"]

In [56]:
# Define a function that converts a dollar amount to millions.
def convert_to_millions(dollar_amount):
    return f"${(dollar_amount / 1000000):,.2f}M"



In [57]:
# Apply the currency_format_millions function to only the money columns. 
for col in money_columns_list:
    top_five_summary_df[col] = top_five_summary_df[col].apply(convert_to_millions)

top_five_summary_df

C:\Users\William Frescas\AppData\Local\Temp\ipykernel_33628\2080775724.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  top_five_summary_df[col] = top_five_summary_df[col].apply(convert_to_millions)


client_id     qty shipping_price total_price total_cost   profit
1      33615   64313         $1.83M      $8.38M     $6.18M   $2.20M
3      46820   75768         $1.60M      $9.74M     $7.01M   $2.74M
4      66037   43018         $1.40M     $10.26M     $7.00M   $3.26M
2      38378   73667         $3.43M     $12.91M     $9.63M   $3.27M
0      24741  239862         $5.13M     $82.27M    $45.69M  $36.58M

In [59]:
# Rename the columns to reflect the change in the money format. 
final_clean_df = top_five_summary_df.rename(columns={
    'client_id': 'Client ID',
    'qty': 'Units',
    'shipping_price': 'Shipping (millions)',
    'total_price': 'Total Revenue (millions)',
    "total_cost": "Total Cost (millions)",
    "profit": "Total Profit (millions)"
})

final_clean_df



Client ID   Units Shipping (millions) Total Revenue (millions)  \
1      33615   64313              $1.83M                   $8.38M   
3      46820   75768              $1.60M                   $9.74M   
4      66037   43018              $1.40M                  $10.26M   
2      38378   73667              $3.43M                  $12.91M   
0      24741  239862              $5.13M                  $82.27M   

  Total Cost (millions) Total Profit (millions)  
1                $6.18M                  $2.20M  
3                $7.01M                  $2.74M  
4                $7.00M                  $3.26M  
2                $9.63M                  $3.27M  
0               $45.69M                 $36.58M

In [61]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
final_clean_df.sort_values("Total Profit (millions)", ascending=False, inplace=True)
final_clean_df

Client ID   Units Shipping (millions) Total Revenue (millions)  \
0      24741  239862              $5.13M                  $82.27M   
2      38378   73667              $3.43M                  $12.91M   
4      66037   43018              $1.40M                  $10.26M   
3      46820   75768              $1.60M                   $9.74M   
1      33615   64313              $1.83M                   $8.38M   

  Total Cost (millions) Total Profit (millions)  
0               $45.69M                 $36.58M  
2                $9.63M                  $3.27M  
4                $7.00M                  $3.26M  
3                $7.01M                  $2.74M  
1                $6.18M                  $2.20M